Doc: access policy fix and expansion (#6573)
edgedb/edgedb: docs/datamodel/access_policies.rst

    }

When no access policies are defined, object-level security is not activated.
Any properly authenticated client can carry out any operation on any object
in the database. At the moment, we would need to ensure that the app itself
handles the logic to restrict users from accessing other users' posts. Access
policies allow us to ensure that the database itself handles this logic,
thereby freeing us up from implementing access control in each and every
piece of software that accesses the data.

.. warning::

    ⚠️ Once a policy is added to a particular object type, **all operations**
    (``select``, ``insert``, ``delete``, ``update``, etc.) on any object of
    that type are now *disallowed by default* unless specifically allowed by an
    access policy! See the subsection on resolution order below for details.

Defining a global
^^^^^^^^^^^^^^^^^

Global variables are the a convenient way to provide the context needed to
determine what sort of access should be allowed for a given object, as they

> can be set and reset as needed.

Author: I think so!

    +     access policy author_has_full_access
    +       allow all
    +       using (global current_user    ?= .author.id
    +         and  global current_country ?= Country.Full) {

>     +        errmessage := "User not in country with full access";

Reviewer: If you do change this, make sure you update the REPL output too.

    +       }
    +      access policy author_has_read_access
    +        allow select
    +        using (global current_user    ?= .author.id 
    +          and  global current_country ?= Country.ReadOnly);
        }

Let's break down the access policy syntax piece-by-piece. These policies grant
full read-write access (``all``) to the ``author`` of each ``BlogPost``, if
the author is in a country that allows full access to the service. Otherwise,
the same author will be restricted to either read-only access or no access at
all, depending on the country.

.. note::

>   We're using the *coalescing equality* operator ``?=`` which returns

Reviewer: Not your change, I know, but while we're here… What do you think?
```suggestion
  We're using the *coalescing equality* operator ``?=`` because it returns
```
This makes it sound less like a side effect and more like the intended effect (which I assume it was).

  ``false`` even if one of its arguments is an empty set.

> - ``access policy`` The keyword used to declare a policy inside an object

Reviewer: I'd vote for restoring the colon here. Maybe it was unintentional?
```suggestion
- ``access policy``: The keyword used to declare a policy inside an object
```

Author: Yes, I think I killed that comma unintentionally -_-

  type.

> - ``author_has_full_access`` and ``author_has_read_access``: The name of these

Reviewer: ```suggestion
- ``author_has_full_access`` and ``author_has_read_access``: The names of these
```

  policies; could be any string.
- ``allow``: The kind of policy; could be ``allow`` or ``deny``
- ``all``: The set of operations being allowed/denied; a comma-separated list

>   of the following: ``all``, ``select``, ``insert``, ``delete``, ``update``,
>   ``update read``, ``update write``.

Reviewer: Making sure it's clear the user doesn't have to list all of these.
```suggestion
  of any number of the following: ``all``, ``select``, ``insert``, ``delete``,
  ``update``, ``update read``, and ``update write``.
```

  db> select count(BlogPost);
  {1}

Next, let's set the ``global current_user`` to some other id. This user has
yet to write any posts, so the number of ``BlogPost`` objects returned via
the ``count`` function is now zero:

.. code-block:: edgeql-repl

  db> set global current_user := <uuid>'d1c64b84-8e3c-11ee-86f0-d7ddecf3e9bd';
  OK: SET GLOBAL
  db> select count(BlogPost);
  {0}

Now let's see what happens when we try to insert a ``BlogPost`` under this new
id. This might represent a situation where our app is misconfigured to allow a
user to switch accounts while still remaining on the blog editing page used
for the previous account. Fortunately, the access policy prevents this post
from being published under the wrong id:

.. code-block:: edgeql-repl

  db> set global current_user := <uuid>'d1c64b84-8e3c-11ee-86f0-d7ddecf3e9bd';
  OK: SET GLOBAL
  db> insert BlogPost {
  ...    title := "My post",
  ...    author := (select User filter .id = global current_user)
  ...  };
  edgedb error: AccessPolicyError: access policy violation on insert of

>   default::BlogPost (User must have same ID as blog author)

Author: Yeah, not a good example as we discussed last week. A better idea struck me which requires two globals but still simple enough and feels much more like a real world example.
